face/unicast-tcp-transport.go

```diff
@@ -174,113 +163,124 @@ func (t *UnicastTCPTransport) GetSendQueueSize() uint64 {
 	return impl.SyscallGetSocketSendQueueSize(rawConn)
 }
 
-// onTransportFailure modifies the state of the UnicastTCPTransport to indicate
-// a failure in transmission.
-func (t *UnicastTCPTransport) onTransportFailure(fromReceive bool) {
-	switch t.persistency {
-	case PersistencyPermanent:
-		// Restart socket
-		t.conn.Close()
-		var err error
-		conn, err := t.dialer.Dial(t.remoteURI.Scheme(), net.JoinHostPort(t.remoteURI.Path(),
-			strconv.Itoa(int(t.remoteURI.Port()))))
-		if err != nil {
-			core.LogError(t, "Unable to connect to remote endpoint: ", err)
-		}
-		t.conn = conn.(*net.TCPConn)
+// Set the connection and params.
+func (t *UnicastTCPTransport) setConn(conn *net.TCPConn) {
+	t.conn = conn
+	t.localAddr = *t.conn.LocalAddr().(*net.TCPAddr)
+	t.localURI = defn.DecodeURIString("tcp://" + t.localAddr.String())
+}
 
-		if fromReceive {
-			t.runReceive()
-		} else {
-			// Old receive thread will error out, so we need to replace it
-			go t.runReceive()
-		}
-	default:
-		t.changeState(defn.Down)
+// Attempt to reconnect to the remote transport.
+func (t *UnicastTCPTransport) reconnect() {
+	// Shut down the existing socket
+	if t.conn != nil {
+		t.conn.Close()
 	}
-}
 
-// expirationHandler checks if the face should expire (if on demand)
-func (t *UnicastTCPTransport) expirationHandler() {
+	// Number of attempts we have made so far
+	attempt := 0
+
+	// Keep trying to reconnect until successful
+	// If the transport is not permanent, do not attempt to restart
+	// Do this inside the loop to account for changes to persistency
 	for {
-		time.Sleep(time.Duration(10) * time.Second)
-		if t.state == defn.Down {
-			break
+		attempt++
+
+		// If there is no connection, this is the initial attempt to
+		// connect for any face, so we will continue regardless
+		// However, make only one attempt to connect for non-permanent faces
+		if !(t.conn == nil && attempt == 1) {
+			// Do not continue if the transport is not permanent or closed
```

Comment:
For non-permanent face, make only one attempt to connect, otherwise drop the face. This is (kinda) similar to the current behavior (which has a different issue - if the connection gets stuck then the forwarder gets stuck completely)